trin-core/src/utp/packets.rs

```diff
@@ -3,13 +3,37 @@ use std::fmt;
 pub const HEADER_SIZE: usize = 20;
 pub const VERSION: u8 = 1;
 
+macro_rules! u8_to_unsigned_be {
+    ($src:ident, $start:expr, $end:expr, $t:ty) => ({
+        (0 .. $end - $start + 1).rev().fold(0, |acc, i| acc | $src[$start+i] as $t << (i * 8))
+    })
+}
+
+macro_rules! make_getter {
+    ($name:ident, $t:ty, $m:ident) => {
+        pub fn $name(&self) -> $t {
+            let header = unsafe { &*(self.0.as_ptr() as *const PacketHeader) };
```

Comment:
I would love to see comments every time we use `unsafe`: what alternatives were considered, why the alternatives were unusable or how much less efficient they would be, etc. Otherwise, it becomes a time-sink, when code readers might wonder if there's a subtle bug lurking in there.

Reply:
In Rust, `unsafe` serves mainly for two purposes: calling unsafe functions and dereferencing raw pointers. 
In this case `*const` and `self.0.as_ptr()` are our raw pointers which are analogous to &T, except that it doesn’t have a lifetime and is not subject to the same validity rule as its & counterpart.

Pointer cast seems a classic Rust pattern and is useful when we want to interpret a sequence
of bytes as #[repr(C)] structs of these. We need this because we don't know the byte length of the packet (self.0) because of the unknown payload, we only know that the first X bytes are the header.